selftests/unit/test_cartesian_config.py

```diff
@@ -409,6 +409,67 @@ def testVariableAssignment(self):
                               ],
                               True)
 
+    def testVariableLazyAssignment(self):
+        self._checkStringDump("""
+            arg1 = ~balabala
+            variants:
+                - base_content:
+                    foo = bar
+                - empty_content:
+            variants:
+                - lazy_set:
+                    foo ~= baz
+                - lazy_set_with_substitution:
+                    foo ~= ${arg1}
+                - dummy_set:
+            foo ~= ~= foo
+            """,
```

Comment:
Is this correctly indented (or visualized here)? The way I understand it, ``foo ~= ~= foo`` is supposed to be restricted to the ``dummy_set`` variant and test for correct handling of the operator within a value.

Reply:
Similar with the `lazy_set` variant tests, the `dummy_set` variant tests wants to verify that the new operator can work properly when it is set at the top level of the configuration, otherwise these two kinds of tests are just the same as each other.

Reply:
Isn't `foo ~= ~= foo` testing if the operator is parsed correctly when used twice (once parsed as operator and once as part of the value? Also, wouldn't setting `foo ~= ~= foo` at the top level of the hierarchy overwrite all combinations? This is what I get from the sample above:

```
dict    1:  lazy_set.base_content
    _name_map_file = {'c.cfg': 'lazy_set.base_content'}
    _short_name_map_file = {'c.cfg': 'lazy_set.base_content'}
    arg1 = ~balabala
    dep = []
    foo = bar
    foo ~ = ~= foo
    name = lazy_set.base_content
    shortname = lazy_set.base_content
dict    2:  lazy_set.empty_content
    _name_map_file = {'c.cfg': 'lazy_set.empty_content'}
    _short_name_map_file = {'c.cfg': 'lazy_set.empty_content'}
    arg1 = ~balabala
    dep = []
    foo ~ = ~= foo
    name = lazy_set.empty_content
    shortname = lazy_set.empty_content
dict    3:  lazy_set_with_substitution.base_content
    _name_map_file = {'c.cfg': 'lazy_set_with_substitution.base_content'}
    _short_name_map_file = {'c.cfg': 'lazy_set_with_substitution.base_content'}
    arg1 = ~balabala
    dep = []
    foo = bar
    foo ~ = ~= foo
    name = lazy_set_with_substitution.base_content
    shortname = lazy_set_with_substitution.base_content
dict    4:  lazy_set_with_substitution.empty_content
    _name_map_file = {'c.cfg': 'lazy_set_with_substitution.empty_content'}
    _short_name_map_file = {'c.cfg': 'lazy_set_with_substitution.empty_content'}
    arg1 = ~balabala
    dep = []
    foo ~ = ~= foo
    name = lazy_set_with_substitution.empty_content
    shortname = lazy_set_with_substitution.empty_content
dict    5:  dummy_set.base_content
    _name_map_file = {'c.cfg': 'dummy_set.base_content'}
    _short_name_map_file = {'c.cfg': 'dummy_set.base_content'}
    arg1 = ~balabala
    dep = []
    foo = bar
    foo ~ = ~= foo
    name = dummy_set.base_content
    shortname = dummy_set.base_content
dict    6:  dummy_set.empty_content
    _name_map_file = {'c.cfg': 'dummy_set.empty_content'}
    _short_name_map_file = {'c.cfg': 'dummy_set.empty_content'}
    arg1 = ~balabala
    dep = []
    foo ~ = ~= foo
    name = dummy_set.empty_content
    shortname = dummy_set.empty_content
```

The way I see it, it is better to test for the double operator syntax in a subvariant and include testing of all other combinations too.

Reply:
Actually I want to test the both within one test. Let's focus on `dummy_set.base_content` and `dummy_set.empty_content` these two variants, the expect results should be:
```
dummy_set.base_content:
    foo = 'bar'
dummy_set.empty_content:
    foo = '~= foo'
```
and the explanation is:

* since `foo` has not been set before the statement `foo ~= ~= foo` being parsed with `dummy_set.empty_content` variant, it is set to `'~= foo'` as what the rules defined
* since `foo` has been set before the statement `foo ~= ~= foo` being parsed with `dummy_set.base_content` variant, its value should still be `'bar'`

Similar with `dummy_set.base_content` variant, `foo` has already been set before the statement `foo ~= ~= foo` being parsed in other tests, so its value should not be changed. And that's the reason of only one `foo = '~= foo'` line should be seen in all tests.

You might mistype the line by adding a whitespace between the first `~=`, that's why there is an abnormal line `foo ~ = ~= foo` in all the variants.

Reply:
Yes, good point about the error - after all the unit test passes too and I might not have checked out at the commit when I ran the above code. I see your points now, I think the confusion for me arose from the fact that you are testing both double use of the operator keyword (`~= ~=`) to detect correctly parsed value (at least I think so from reading the changes) and top level overwriting in the same round. Using a clearer separation you could say:

```
arg1 = ~balabala
variants:
    - base_content:
        foo = bar
    - empty_content:
variants:
    - lazy_set:
        foo ~= baz
    - lazy_set_with_substitution:
        foo ~= ${arg1}
    - lazy_set_with_double_token:
        foo ~= ~= foo
    - dummy_set:
foo ~= baz
```

but this is not really important. Just wanted to share perspective from somebody reading the unit test and trying to guess what it is testing against here. Feel free to resolve this too.

Reply:
Good suggestion, will update accordingly.